### ISSUE TITLE
Exponentially bad performance in minifyJS (GHC 7.10.3)

### ISSUE THREAD

**Issue by paul-rouse** [erikd/language-javascript#58]
This is quite odd, because it seems to happen with GHC 7.10.3 (stackage lts-6.23) but goes away with GHC 8.0.1 (stackage nightly).  Obviously a few dependency versions change too.  *This is a mistake: the bug is there with GHC 8 too.  I must have mixed up executables for different versions of the test, or something*

In language-javascript-0.6.0.8 the time taken by minifyJS can be exponential in the length of an expression.  For example, when trying to minify "var x = 'a'+1+'a'+1+'a'+1+'a';" the time goes up by a factor of four for each additional "'a'+1+'".

This was discovered when looking into https://github.com/yesodweb/yesod/issues/1291

FWIW here is my complete test program

```haskell
import System.Environment                 (getArgs)
import Language.JavaScript.Parser         (parse)
import Language.JavaScript.Process.Minify (minifyJS)

-- Give a single integer on the command line to specify the number of
-- copies of "'a'+1' in the expression.  But be careful: the time taken
-- for minifyJS goes up as 4^n !!
main :: IO ()
main = do
    args <- getArgs
    let n = case args of
                nn:_ -> read nn :: Int
                _    -> error "Specify repetition count on the command line"
        expr = "var x = " ++ concat (replicate n "'a'+1+") ++ "'a';"
    putStrLn $ "Minifying: " ++ expr
    let x = parse expr "noname"
    case x of
        Left err  -> putStrLn $ "This shouldn't happen: " ++ err
        Right ast -> putStrLn $ show $ minifyJS ast
```

**Comment by erikd** [comment]
Are you testing the save version of language-javascript with the two compilers?

**Comment by paul-rouse** [comment]
Yes, 0.6.0.8 in both cases.

**Comment by erikd** [comment]
Ok, I have two executables from your program above, one built with ghc-7.10.3 and one with ghc-8.0.1. In both cases I compiled with `ghc -Wall -O2 perf-test.hs -o perf-test-XX` and they give pretty much identical results:
```
> time ./perf-test-7.10.3 15 > /dev/null

real	0m22.585s
user	0m22.428s
sys	0m0.112s

> time ./perf-test-8.0.1 15 > /dev/null

real	0m22.477s
user	0m22.356s
sys	0m0.100s
```
I agree, those results are far from great. Disappointingly replacing `show` with `renderToString` (which obviously prints the minified versions rather than the AST) is no better (and ghc-8.0.1 is marginally better than ghc-7.10.3). Switching to `renderToText` doesn't seem to help either.

So, I cannot reproduce any significant difference in performance between 7.10.3 and 8.0.1 I can see that performance is exponential.

**Comment by erikd** [comment]
The expression we're minifying here is definitely pathological. The AST with the `JSAnnot`s removed look like:
```
JSExpressionBinary
	(JSExpressionBinary
		(JSExpressionBinary
			(JSExpressionBinary (JSStringLiteral  "'a'") JSBinOpPlus (JSDecimal  "1"))
            JSBinOpPlus (JSStringLiteral  "'a'"))
        JSBinOpPlus (JSDecimal  "1"))
    JSBinOpPlus (JSStringLiteral  "'a'")
```
Obviously, I could probably add a fix for this particular pathological case, but I'm sure there are lots of others so I'm somewhat hesitant to fix just this one.

**Comment by erikd** [comment]
Tried just printing the AST without minification and that is almost instantaneous, so its obviously minification that's causing the performance problems.

**Comment by erikd** [comment]
The exponential performance of minification is almost certainly due to the `JSExpressionBinary` could be a balanced tree but isn't.

**Comment by erikd** [comment]
Simple test program that generates the same logical expression, once as a balanced binary tree and the other as a wildly right biased tree (as generated by the parser). When minifying these two the balanced one is lightning fast, whereas the unbalanced one is *really* slow.
```
import System.Environment
import Language.JavaScript.Parser
import Language.JavaScript.Parser.AST
import Language.JavaScript.Process.Minify
import System.IO
import Data.List

-- Give a single integer on the command line to specify the number of
-- copies of "'a'+1' in the expression.  But be careful: the time taken
-- for minifyJS goes up as 4^n !!
main :: IO ()
main = do
    args <- getArgs
    let n = case args of
                nn:_ -> read nn :: Int
                _    -> error "Specify repetition count on the command line"

    runTest $ balancedTest (2 ^ n)
    runTest $ unbalancedTest (2 ^ n)

runTest :: JSAST -> IO ()
runTest expr = do
    hFlush stdout
    print $ renderToString expr
    print $ minifyJS expr
    putStrLn ""
    hFlush stdout


unbalancedTest :: Int -> JSAST
unbalancedTest n =
    let expr = "var y = " ++ intercalate " + " (replicate (n `div` 2) "'a' + 1") ++ " ; " in
    case parse expr "noname" of
        Left err -> error $ "This shouldn't happen: " ++ err
        Right ast -> ast

balancedTest :: Int -> JSAST
balancedTest count = JSAstProgram
    [ JSVariable JSAnnotSpace (JSLOne
        ( JSVarInitExpression
            ( JSIdentifier JSAnnotSpace "y")
            ( JSVarInit JSAnnotSpace (balancedPlusOne $ count - 1))
            ))
      (JSSemi JSAnnotSpace)
    ] JSAnnotSpace
  where
    aPlusOne :: JSExpression
    aPlusOne = JSExpressionBinary
                (JSStringLiteral JSAnnotSpace "'a'")
                (JSBinOpPlus JSAnnotSpace)
                (JSDecimal JSAnnotSpace "1")

    balancedPlusOne :: Int -> JSExpression
    balancedPlusOne n
        | n <= 2 = aPlusOne
        | otherwise = JSExpressionBinary
                        (balancedPlusOne $ n `div` 2)
                        (JSBinOpPlus JSAnnotSpace)
                        (balancedPlusOne $ n `div` 2)
```
I've bee compiling this in a sandbox with `cabal exec -- ghc -Wall -O2 perf-test2.hs -o perf-test2` and running it as `./perf-test2 5`.

**Comment by paul-rouse** [comment]
While I can see the tree is unbalanced, there must be something wrong if the time for minification is exponential in the depth of the tree!  

I agree that the expression I wrote looks pretty artificial, but I think it is important to remember that it arose as I was trying to reduce a real example to something minimal.  The original code which the user had trouble with is shown in the Yesod issue (https://github.com/yesodweb/yesod/issues/1291), and it looks perfectly reasonable to me.

As for GHC 8, maybe I fooled myself somehow - I'll try to work it out - the fact that you see the bug is all that matters really!

**Comment by paul-rouse** [comment]
I've just been having a (very quick) look at your code.  Isn't there something wrong in the relationship between [the JSExpressionBinary case of *fix*](https://github.com/erikd/language-javascript/blob/new-ast/src/Language/JavaScript/Process/Minify.hs#L148) and [the general case of *fixBinOpExpression*](https://github.com/erikd/language-javascript/blob/new-ast/src/Language/JavaScript/Process/Minify.hs#L171)?  They **both** seem to recurse down both lhs and rhs!  That looks as if it would explain doubling the amount of work with each level of the tree.

Do you need the recursion at all at [line 148](https://github.com/erikd/language-javascript/blob/new-ast/src/Language/JavaScript/Process/Minify.hs#L148), given that it is done in *fixBinOpExpression* anyway? Or have I missed something?

BTW This explains why I needed to alternate strings and numbers - if all operands are litereal strings the optimisation in the first case of *fixBinOpExpression* is invoked, avoiding the double recursion.

**Comment by erikd** [comment]
The double recursion may be part of the problem, but even after fixing that, minification of the unbalanced tree is still significantly slower than for a balanced tree with the same number of leaf nodes.

**Comment by paul-rouse** [comment]
I think the string concatenation case of *fixBinOpExpression* needs to be rewritten to do its own recursion, if the recursion in line 148 is removed.  Or maybe the recursion needs to be left in line 148 and removed completely from all cases of *fixBinOpExpression* - I don't really understand how the JSAnnot values are being used, since they seem to be slightly different in the two recursions.

The unbalanced tree may be slightly slower to process - although I don't think it should be - but it would be totally unreasonable to expect users to write their javascript so that all expressions parsed to balanced trees!  Are you suggesting automatically balancing the tree?  That seems harder to me than removing the double recursion.

**Comment by paul-rouse** [comment]
To expand a bit on my first point just now - if the recursion is removed from line 148, then I think line 168 would need expanding roughly like this.  Except that I haven't looked very hard at how the `a` value is used.  I think the other cases of *fixBinOpExpression* already do the recursions they need.

```haskell
fixBinOpExpression a (JSBinOpPlus x) lhs rhs =
    let lhs' = fix a lhs
        rhs' = fixEmpty rhs
    in case (lhs', rhs') of
        ((JSStringLiteral _ s1), (JSStringLiteral _ s2)) -> stringLitConcat s1 s2
        _ -> JSExpressionBinary lhs' (fixEmpty $ JSBinOpPlus x) rhs'
```

**Comment by erikd** [comment]
@paul-rouse Pretty sure I've fixed this issue in the latest set of commits. If you can confirm that I will release a new version to Hackage.

**Comment by paul-rouse** [comment]
The solution I gave in my previous comment also works, and is much simpler.  It really is the double recursion, not the unbalanced tree.  But it's your choice.

**Comment by erikd** [comment]
You are correct. Fixed to remove double recursion and 0.6.0.9 uploaded to Hackage.

**Comment by paul-rouse** [comment]
Thank you for spending time exploring this!
